src/WebJobs.Script/Host/FunctionMetadataManager.cs

```diff
@@ -285,5 +294,53 @@ private void AddMetadataFromCustomProviders(IEnumerable<IFunctionProvider> funct
                 }
             }
         }
+
+        private void ValidateHostJsonFile()
+        {
+            try
+            {
+                // Search for the host.json file within nested directories to verify scenarios where it isn't located at the root. This situation often occurs when a function app has been improperly zipped.
+                string hostFilePath = Path.Combine(_scriptOptions.Value.RootScriptPath, ScriptConstants.HostMetadataFileName);
+                IEnumerable<string> hostJsonFiles = Directory.GetFiles(_scriptOptions.Value.RootScriptPath, ScriptConstants.HostMetadataFileName, SearchOption.AllDirectories)
+                    .Where(file => !file.Equals(hostFilePath, StringComparison.OrdinalIgnoreCase));
+
+                if (IsDefaultHostConfig())
+                {
+                    if (hostJsonFiles != null && hostJsonFiles.Any())
+                    {
+                        string hostJsonFilesPath = string.Join(", ", hostJsonFiles).Replace(_scriptOptions.Value.RootScriptPath, string.Empty);
+                        _logger.HostJsonZipDeploymentIssue(hostJsonFilesPath);
+                    }
+                    else
+                    {
+                        _logger.NoHostJsonFile();
+                    }
+                }
+                else
+                {
+                    _logger.MissingFunctionsDeploymentIssue();
+                }
+            }
+            catch
+            {
+                // Ignore any exceptions.
+            }
+        }
+
+        private bool IsDefaultHostConfig()
+        {
+            IConfiguration configuration = _serviceProvider?.GetRequiredService<IConfiguration>();
```

Comment:
Can we try to read this value from `ScriptApplicationHostOptions`? Would be good to avoid manually reading from IConfiguration .

Reply:
+1

Reply:
Done.